### PR TITLE
Support Tigris pushdowns for numbers

### LINKED ISSUE #1755: Support pushdowns for equal operators on numbers
### What should be done?

Currently, we only pushdown queries where `_id` is equal to any `ObjectId` or `string`.
We should generalize pushdowns to all keys and most importantly - initialize pushdowns for a simple number comparison. 
- We should only pushdown queries where we check if numbers from all numeric types are equal (other comparison types are not in a scope)
- After we introduce pushdown tests in #1279, there shouldn't be any problem with testing.
- We shouldn't pushdown number comparison with any non numerical type, although we need to make sure that we don't omit any data by not including it in pushdown query response because of different types
- don't recurse yet, we shouldn't check for array items and subdocuments. I guess we could exclude every filter which tries to compare in array/subdocument, but please create test cases for that
 
## DoD
- Pushdown filters that check if a field is equal to any number from any numeric type
- Update and create compat tests to include enough edge cases related to type conversions (e.g. using fields of different types)
- Focus on tests related to comparing nonnumerical types with numerical to make sure that we don't leave any data (unit tests might be also a good idea)
- Implement the straightforward way, no type conversion or complicate cases. If we find that type conversion is needed or some cases work on Postgres/Tigris, create separate issue(-s)
- If without conversion noting is possible, we consider the current task a research task and create separate issues based on the research result

### PR DESCRIPTION
**Pull request by noisersup**
# Description

Closes #1755.

## Readiness checklist

* [ ] I added unit tests for new functionality or bug fixes.
* [ ] I added integration tests for new functionality or bug fixes.
* [x] I added compatibility tests for new functionality or bug fixes.
* [x] I made spot refactorings.
* [ ] I updated user documentation.
* [ ] I ran `task all`, and it passed.
* [ ] I added/updated comments for both exported and unexported top-level declarations (functions, types, etc).
* [ ] I checked comments rendering with `task godocs`.
* [x] I ensured that the title is good enough for the changelog.
* [x] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Assignee, Labels, Project and project's Sprint fields.
* [x] I marked all done items in this checklist.

### CODE REVIEW

**Comment by noisersup** [comment]
This PR is halted, the current state is that it works for most cases, but there are a few edge cases as the fact that filtering the `""` doesn't return the empty strings (maybe the tigris bug?), or float comparisons on big numbers (though we don't expect the float comparison will work correctly)

File: internal/handlers/tigris/tigrisdb/query.go [hunk 1]
```diff
@@ -95,25 +110,45 @@ func (tdb *TigrisDB) BuildFilter(filter *types.Document) driver.Filter {
 	res := map[string]any{}
 
 	for k, v := range filter.Map() {
-		// filter only by _id for now
-		if k != "_id" {
+		// don't pushdown empty strings yet
```
Comment (noisersup) [inline internal/handlers/tigris/tigrisdb/query.go:113]: For those cases, I believe that the empty string filter is the last one in terms of incompatibility. They work currently on API, but not on driver. I assume it'll be shortly fixed so I'll cover that on dotnotation task.

Comment (rumyantseva) [reply internal/handlers/tigris/tigrisdb/query.go:113]: @noisersup let's create a separate issue and add a TODO here?

File: internal/handlers/tigris/tigrisdb/query.go [hunk 2]
```diff
@@ -95,25 +110,45 @@ func (tdb *TigrisDB) BuildFilter(filter *types.Document) driver.Filter {
...
-		switch v.(type) {
-		case string:
-			// filtering by string values is complicated if the storage supports encodings, collations, etc,
-			// but Tigris does not support any of these
-		case types.ObjectID:
-			// filtering by ObjectID is always safe
-		default:
-			// skip other types for now
+		if k != "" {
+			// don't pushdown $comment and dot notation yet
```
Comment (rumyantseva) [inline internal/handlers/tigris/tigrisdb/query.go:119]: Let's add TODO and an issue number

File: internal/handlers/tigris/tigrisdb/query.go [hunk 3]
```diff
@@ -95,25 +110,45 @@ func (tdb *TigrisDB) BuildFilter(filter *types.Document) driver.Filter {
...
+				continue
+			}
+		}
+
+		// _id field supports only specific types
+		if k == "_id" {
+			switch v.(type) {
+			case *types.Document, *types.Array, types.Binary, bool, time.Time, types.NullType, types.Regex, types.Timestamp:
+			// type not supported for pushdown
+
```
Comment (rumyantseva) [inline internal/handlers/tigris/tigrisdb/query.go:130]: Formatting in this switch-case looks a bit insocnistent to me: 
- Should each `case` be ended with an empty line or not?
- Should a comment be shifter in this case (like in the next one)?

In the switch below, the same things are formated differently.

File: internal/handlers/tigris/tigrisdb/query.go [hunk 4]
```diff
@@ -95,25 +110,45 @@ func (tdb *TigrisDB) BuildFilter(filter *types.Document) driver.Filter {
...
+				panic(fmt.Sprintf("Unexpected type of field %s: %T", k, v))
+			}
+
 			continue
 		}
 
-		// filter by the exact _id value
-		id := must.NotFail(tjson.Marshal(v))
-		res["_id"] = json.RawMessage(id)
+		switch v.(type) {
```
Comment (rumyantseva) [inline internal/handlers/tigris/tigrisdb/query.go:142]: What is the difference between `k == "_id"` and other fields? Both switches look quite identical to me.

**Review by rumyantseva — COMMENTED** [review]
Great job overall! I asked a few questions.

File: internal/handlers/tigris/tigrisdb/query.go (hunk 4 above)
Comment (rumyantseva) [reply internal/handlers/tigris/tigrisdb/query.go:142]: Well, at the moment, we don't validate queries. If someone tries to set `_id` as an array, what will be the result? Nothing will be found, and it's an expected result. Why should we validate queries then?

Comment (noisersup) [reply internal/handlers/tigris/tigrisdb/query.go:142]: valid point, going to fix that

File: internal/handlers/tigris/tigrisdb/query.go (hunk 3 above)
Comment (chilagrow) [reply internal/handlers/tigris/tigrisdb/query.go]: > Should each case be ended with an empty line or not?

I think if we agree we can put that logic into our linter :hugs:

**Review by chilagrow — APPROVED** [review]

File: internal/handlers/tigris/tigrisdb/query.go [hunk 5]
```diff
@@ -85,7 +88,19 @@ func (tdb *TigrisDB) QueryDocuments(ctx context.Context, param *FetchParam) ([]*
...
 
-	return res, iter.Err()
+	err = iter.Err()
+
+	switch {
+	case err == nil:
+		fallthrough
+	case IsInvalidArgument(err):
+		// Skip errors from filtering invalid types
+		break
```
Comment (AlekSi) [inline internal/handlers/tigris/tigrisdb/query.go:98]: @noisersup Why do we do that? IIUC, we just hide what could be a real error

Comment (noisersup) [reply internal/handlers/tigris/tigrisdb/query.go:98]: InvalidArgument error occures while filtering the schema with wrong type. For example if `v` is a string, but we use filter `{"v":42}` we get this error. MongoDB on this scenario result with no error, as it just means that filter don't match the value in database

Comment (AlekSi) [reply internal/handlers/tigris/tigrisdb/query.go:98]: ic, thanks
